README
NAME
    `Net::Async::FastCGI' - use FastCGI with IO::Async

SYNOPSIS
    As an adapter:

     use Net::Async::FastCGI;
     use IO::Async::Loop;

     my $loop = IO::Async::Loop->new();

     my $fastcgi = Net::Async::FastCGI->new(
        on_request => sub {
           my ( $fastcgi, $req ) = @_;

           # Handle the request here
        }
     );

     $loop->add( $fastcgi );

     $fastcgi->listen(
        service => 1234,
        on_resolve_error => sub { die "Cannot resolve - $_[-1]\n" },
        on_listen_error  => sub { die "Cannot listen - $_[-1]\n" },
     );

     $loop->loop_forever;

    As a subclass:

     package MyFastCGIResponder;
     use base qw( Net::Async::FastCGI );

     sub on_request
     {
        my $self = shift;
        my ( $req ) = @_;

        # Handle the request here
     }

     ...

     use IO::Async::Loop;

     my $loop = IO::Async::Loop->new();

     my $fastcgi;
     $loop->add( $fastcgi = MyFastCGIResponder->new( service => 1234 ) );

     $fastcgi->listen(
        service => 1234,
        on_resolve_error => sub { die "Cannot resolve - $_[-1]\n" },
        on_listen_error  => sub { die "Cannot listen - $_[-1]\n" },
     );

     $loop->loop_forever;

DESCRIPTION
    This module allows a program to respond asynchronously to FastCGI
    requests, as part of a program based on IO::Async. An object in this
    class represents a single FastCGI responder that the webserver is
    configured to communicate with. It can handle multiple outstanding
    requests at a time, responding to each as data is provided by the
    program. Individual outstanding requests that have been started but not
    yet finished, are represented by instances of
    Net::Async::FastCGI::Request.

EVENTS
    The following events are invoked, either using subclass methods or CODE
    references in parameters:

  on_request $req
    Invoked when a new FastCGI request is received. It will be passed a new
    Net::Async::FastCGI::Request object.

PARAMETERS
    The following named parameters may be passed to `new' or `configure':

    on_request => CODE
            CODE references for `on_request' event handler.

    default_encoding => STRING
            Sets the default encoding used by all new requests. If not
            supplied then `UTF-8' will apply.

METHODS
  $fcgi->listen( %args )
    Start listening for connections on a socket, creating it first if
    necessary.

    This method may be called in either of the following ways. To listen on
    an existing socket filehandle:

    handle => IO
        An IO handle referring to a listen-mode socket. This is now
        deprecated; use the `handle' key to the `new' or `configure' methods
        instead.

    Or, to create the listening socket or sockets:

    service => STRING
        Port number or service name to listen on.

    host => STRING
        Optional. If supplied, the hostname will be resolved into a set of
        addresses, and one listening socket will be created for each
        address. If not, then all available addresses will be used.

    This method may also require `on_listen_error' or `on_resolve_error'
    callbacks for error handling - see IO::Async::Listener for more detail.

Limits in FCGI_GET_VALUES
    The `FCGI_GET_VALUES' FastCGI request can enquire of the responder the
    maximum number of connections or requests it can support. Because this
    module puts no fundamental limit on these values, it will return some
    arbitrary numbers. These are given in package variables:

     $Net::Async::FastCGI::MAX_CONNS = 1024;
     $Net::Async::FastCGI::MAX_REQS  = 1024;

    These variables are provided in case the containing application wishes
    to make the library return different values in the request. These values
    are not actually used by the library, other than to fill in the values
    in response of `FCGI_GET_VALUES'.

Using a socket on STDIN
    When running a local FastCGI responder, the webserver will create a new
    INET socket connected to the script's STDIN file handle. To use the
    socket in this case, it should be passed as the `handle' argument.

SEE ALSO
    *   CGI::Fast - Fast CGI drop-in replacement of CGI; single-threaded,
        blocking mode.

    *   http://hoohoo.ncsa.uiuc.edu/cgi/interface.html - The Common Gateway
        Interface Specification

    *   http://www.fastcgi.com/devkit/doc/fcgi-spec.html - FastCGI
        Specification

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Net-Async-FastCGI

0.24    CHANGES:
         * Correct SYNOPSIS examples to use ->listen rather than 'service'
           param
         * Import Exporter's 'import' function rather than 'use base'ing on
           Exporter

0.23    CHANGES:
         * Implement on_request as a normal IO::Async::Notifier-style event,
           so it works as a subclass too

0.22    CHANGES:
         * Split out from FCGI-Async; most code renamed
         * Net::Async::FastCGI::Protocol directly built on top of
           IO::Async::Protocol::Stream rather than overriding ::Stream
           directly and replacing some methods

Revision history for FCGI-Async

0.21    CHANGES:
         * Implement HTTP::Request/HTTP::Response gatewaying
         * Implement PSGI gatewaying
         * Provide Plack::Handler::FCGI::Async for plackup et.al.
         * Implement ->stdin, ->stdout, ->stderr pseudo-filehandles

0.20    CHANGES:
         * Use Net::FastCGI for low-level FastCGI constants and message
           handling functions
         * Ensure that ->set_encoding( undef ) actually works to disable the
           encoding mechanism

        BUGFIXES:
         * Collect the entire PARAMS stream and only parse it when it's all
           present, rather than piecewise

0.19    CHANGES:
         * import Exporter::import instead of @ISAing it
         * Respond with FCGI_UNKNOWN_TYPE or FCGI_UNKNOWN_ROLE when
           appropriate (fixes RT 54480)
         * Updated bundled example applications to a modern IO::Async style

0.18    CHANGES:
         * Allow setting of per-request encoding for STDIN/STDOUT/STDERR
           streams
         * Code adjustment to make use of the new IO::Async::Listener class
         * Deprecated ->listen( handle => IO ) in favour of ->new or
           ->configure instead.

0.17    CHANGES:
         * Added 'use warnings'
         * Documentation updates
         * Various small updates to keep CPANTS happy

0.16    BUGFIXES:
         * Support FCGI_GET_VALUES 
           (closes http://rt.cpan.org/Ticket/Display.html?id=43976)
         * Use Test::HexString and wait_for_stream() during testing

0.15    BUGFIXES:
         * Correctly handle webserver-aborted requests - silently discard
           output.

0.14    CHANGES:
         * Reworked constructor to use IO::Async::Loop->listen(). Allows
           specifying a specific hostname.
         * Added Request->stream_stdout_then_finish().
         * Combine small stream writes into larger ones, to gain overhead
           efficiences over the TCP socket.

        BUGFIXES:
         * Respect the FCGI_KEEP_CONN flag to close connections if required

0.13    CHANGES:
         * Updated for IO::Async 0.11:
            + IO::Async::Set is now ::Loop
            + IO::Async::Buffer is now ::Stream
            + Use of $loop->watch_child() in examples rather than hand-coded
              around watching SIGCHLD directly.

0.12    BUGFIXES:
         * Updated to IO::Async::Buffer 0.10 (method/event renames)

0.11    CHANGES:
         * Allow Request->finish() to take an exitcode

        BUGFIXES:
         * Cope with environment parameters longer than 127 bytes
         * Small updates to included 'example' scripts

0.10    CHANGES:
         * Added CGI->FastCGI gateway example

        BUGFIXES:
         * Better handling of ->read_stdin_line()

0.09:   CHANGES:
         * Added 'fortune' example
         * Better testing of connection reuse
         * Support printing to STDERR FastCGI stream

0.08:   CHANGES:
         * First version to be based on IO::Async

0.07:   CHANGES:
         * Changed build system from ExtUtils::MakeMaker to Module::Build

Versions before this did not appear on CPAN, and no 'Changes' notes are
provided for them.


lib/Net/Async/FastCGI.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2005-2011 -- leonerd@leonerd.org.uk

package Net::Async::FastCGI;

use strict;
use warnings;

use Carp;

use base qw( IO::Async::Listener );
IO::Async::Listener->VERSION( '0.35' );

use Net::Async::FastCGI::ServerProtocol;

use IO::Socket::INET;

our $VERSION = '0.24';

# The FCGI_GET_VALUES request might ask for our maximally supported number of
# concurrent connections or requests. We don't really have an inbuilt maximum,
# so just respond these large numbers
our $MAX_CONNS = 1024;
our $MAX_REQS  = 1024;

=head1 NAME

C<Net::Async::FastCGI> - use FastCGI with L<IO::Async>

=head1 SYNOPSIS

As an adapter:

 use Net::Async::FastCGI;
 use IO::Async::Loop;

 my $loop = IO::Async::Loop->new();

 my $fastcgi = Net::Async::FastCGI->new(
    on_request => sub {
       my ( $fastcgi, $req ) = @_;

       # Handle the request here
    }
 );

 $loop->add( $fastcgi );

 $fastcgi->listen(
    service => 1234,
    on_resolve_error => sub { die "Cannot resolve - $_[-1]\n" },
    on_listen_error  => sub { die "Cannot listen - $_[-1]\n" },
 );

 $loop->loop_forever;

As a subclass:

 package MyFastCGIResponder;
 use base qw( Net::Async::FastCGI );

 sub on_request
 {
    my $self = shift;
    my ( $req ) = @_;

    # Handle the request here
 }

 ...

 use IO::Async::Loop;

 my $loop = IO::Async::Loop->new();

 my $fastcgi;
 $loop->add( $fastcgi = MyFastCGIResponder->new( service => 1234 ) );

 $fastcgi->listen(
    service => 1234,
    on_resolve_error => sub { die "Cannot resolve - $_[-1]\n" },
    on_listen_error  => sub { die "Cannot listen - $_[-1]\n" },
 );

 $loop->loop_forever;

=head1 DESCRIPTION

This module allows a program to respond asynchronously to FastCGI requests,
as part of a program based on L<IO::Async>. An object in this class represents
a single FastCGI responder that the webserver is configured to communicate
with. It can handle multiple outstanding requests at a time, responding to
each as data is provided by the program. Individual outstanding requests that
have been started but not yet finished, are represented by instances of
L<Net::Async::FastCGI::Request>.

=cut

=head1 EVENTS

The following events are invoked, either using subclass methods or CODE
references in parameters:

=head2 on_request $req

Invoked when a new FastCGI request is received. It will be passed a new
L<Net::Async::FastCGI::Request> object.

=cut

=head1 PARAMETERS

The following named parameters may be passed to C<new> or C<configure>:

=over 8

=item on_request => CODE

CODE references for C<on_request> event handler.

=item default_encoding => STRING

Sets the default encoding used by all new requests. If not supplied then
C<UTF-8> will apply.

=back

=cut

sub _init
{
   my $self = shift;
   my ( $params ) = @_;
   $self->SUPER::_init( $params );

   $params->{default_encoding} = "UTF-8";
}

sub configure
{
   my $self = shift;
   my %params = @_;

   if( exists $params{on_request} ) {
      $self->{on_request} = delete $params{on_request};
   }

   if( exists $params{default_encoding} ) {
      $self->{default_encoding} = delete $params{default_encoding};
   }

   $self->SUPER::configure( %params );
}

sub on_stream
{
   my $self = shift;
   my ( $stream ) = @_;

   $self->add_child( Net::Async::FastCGI::ServerProtocol->new(
      transport => $stream,
      fcgi      => $self,
   ) );
}

=head1 METHODS

=cut

=head2 $fcgi->listen( %args )

Start listening for connections on a socket, creating it first if necessary.

This method may be called in either of the following ways. To listen on an
existing socket filehandle:

=over 4

=item handle => IO

An IO handle referring to a listen-mode socket. This is now deprecated; use
the C<handle> key to the C<new> or C<configure> methods instead.

=back

Or, to create the listening socket or sockets:

=over 4

=item service => STRING

Port number or service name to listen on.

=item host => STRING

Optional. If supplied, the hostname will be resolved into a set of addresses,
and one listening socket will be created for each address. If not, then all
available addresses will be used.

=back

This method may also require C<on_listen_error> or C<on_resolve_error>
callbacks for error handling - see L<IO::Async::Listener> for more detail.

=cut

sub listen
{
   my $self = shift;
   my %args = @_;

   if( $args{handle} ) {
      carp "Using 'handle' as a ->listen argument is deprecated; use the ->configure method instead";
      $self->configure( handle => $args{handle} );
   }
   else {
      $self->SUPER::listen( %args, socktype => SOCK_STREAM );
   }
}

sub _request_ready
{
   my $self = shift;
   my ( $req ) = @_;

   $self->invoke_event( on_request => $req );
}

sub _default_encoding
{
   my $self = shift;
   return $self->{default_encoding};
}

=head1 Limits in FCGI_GET_VALUES

The C<FCGI_GET_VALUES> FastCGI request can enquire of the responder the
maximum number of connections or requests it can support. Because this module
puts no fundamental limit on these values, it will return some arbitrary
numbers. These are given in package variables:

 $Net::Async::FastCGI::MAX_CONNS = 1024;
 $Net::Async::FastCGI::MAX_REQS  = 1024;

These variables are provided in case the containing application wishes to make
the library return different values in the request. These values are not
actually used by the library, other than to fill in the values in response of
C<FCGI_GET_VALUES>.

=head1 Using a socket on STDIN

When running a local FastCGI responder, the webserver will create a new INET
socket connected to the script's STDIN file handle. To use the socket in this
case, it should be passed as the C<handle> argument.

=head1 SEE ALSO

=over 4

=item *

L<CGI::Fast> - Fast CGI drop-in replacement of L<CGI>; single-threaded,
blocking mode.

=item *

L<http://hoohoo.ncsa.uiuc.edu/cgi/interface.html> - The Common Gateway
Interface Specification

=item *

L<http://www.fastcgi.com/devkit/doc/fcgi-spec.html> - FastCGI Specification

=back

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/Net/Async/FastCGI/Request.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2005-2011 -- leonerd@leonerd.org.uk

package Net::Async::FastCGI::Request;

use strict;
use warnings;

use Carp;

use Net::FastCGI::Constant qw( :type :flag :protocol_status );
use Net::FastCGI::Protocol qw( 
   parse_params
   build_end_request_body
);

# The largest amount of data we can fit in a FastCGI record - MUST NOT
# be greater than 2^16-1
use constant MAXRECORDDATA => 65535;

use Encode qw( find_encoding );
use POSIX qw( EAGAIN );

our $VERSION = '0.24';

my $CRLF = "\x0d\x0a";

=head1 NAME

C<Net::Async::FastCGI::Request> - a single active FastCGI request

=head1 SYNOPSIS

 use Net::Async::FastCGI;
 use IO::Async::Loop;

 my $fcgi = Net::Async::FastCGI->new(
    on_request => sub {
       my ( $fcgi, $req ) = @_;

       my $path = $req->param( "PATH_INFO" );
       $req->print_stdout( "Status: 200 OK\r\n" .
                           "Content-type: text/plain\r\n" .
                           "\r\n" .
                           "You requested $path" );
       $req->finish();
    }
 );

 my $loop = IO::Async::Loop->new();

 $loop->add( $fcgi );

 $loop->loop_forever;

=head1 DESCRIPTION

Instances of this object class represent individual requests received from the
webserver that are currently in-progress, and have not yet been completed.
When given to the controlling program, each request will already have its
parameters and STDIN data. The program can then write response data to the
STDOUT stream, messages to the STDERR stream, and eventually finish it.

This module would not be used directly by a program using
C<Net::Async::FastCGI>, but rather, objects in this class are passed into the
C<on_request> callback of the containing C<Net::Async::FastCGI> object.

=cut

sub new
{
   my $class = shift;
   my %args = @_;

   my $rec = $args{rec};

   my $self = bless {
      conn       => $args{conn},
      fcgi       => $args{fcgi},

      reqid      => $rec->{reqid},
      keepconn   => $rec->{flags} & FCGI_KEEP_CONN,

      stdin      => "",
      stdindone  => 0,
      params     => {},
      paramsdone => 0,

      stdout     => "",
      stderr     => "",

      used_stderr => 0,
   }, $class;

   $self->set_encoding( $args{fcgi}->_default_encoding );

   return $self;
}

sub write_record
{
   my $self = shift;
   my ( $rec ) = @_;

   return if $self->is_aborted;

   my $content = $rec->{content};
   my $contentlen = length( $content );
   if( $contentlen > MAXRECORDDATA ) {
      warn __PACKAGE__."->write_record() called with content longer than ".MAXRECORDDATA." bytes - truncating";
      $content = substr( $content, 0, MAXRECORDDATA );
   }

   $rec->{reqid} = $self->{reqid} unless defined $rec->{reqid};

   my $conn = $self->{conn};

   $conn->write_record( $rec, $content );

}

sub incomingrecord
{
   my $self = shift;
   my ( $rec ) = @_;

   my $type    = $rec->{type};

   if( $type == FCGI_PARAMS ) {
      $self->incomingrecord_params( $rec );
   }
   elsif( $type == FCGI_STDIN ) {
      $self->incomingrecord_stdin( $rec );
   }
   else {
      warn "$self just received unknown record type";
   }
}

sub _ready_check
{
   my $self = shift;

   if( $self->{stdindone} and $self->{paramsdone} ) {
      $self->{fcgi}->_request_ready( $self );
   }
}

sub incomingrecord_params
{
   my $self = shift;
   my ( $rec ) = @_;

   my $content = $rec->{content};
   my $len     = $rec->{len};

   if( $len ) {
      no warnings 'uninitialized';
      $self->{paramscontent} .= $content;
      return;
   }
   else {
      $self->{params} = parse_params( delete $self->{paramscontent} );
      $self->{paramsdone} = 1;
   }

   $self->_ready_check;
}

sub incomingrecord_stdin
{
   my $self = shift;
   my ( $rec ) = @_;

   my $content = $rec->{content};
   my $len     = $rec->{len};

   if( $len ) {
      $self->{stdin} .= $content;
   }
   else {
      $self->{stdindone} = 1;
   }

   $self->_ready_check;
}

=head1 METHODS

=cut

=head2 $hashref = $req->params

This method returns a reference to a hash containing a copy of the request
parameters that had been sent by the webserver as part of the request.

=cut

sub params
{
   my $self = shift;

   my %p = %{$self->{params}};

   return \%p;
}

=head2 $p = $req->param( $key )

This method returns the value of a single request parameter, or C<undef> if no
such key exists.

=cut

sub param
{
   my $self = shift;
   my ( $key ) = @_;

   return $self->{params}{$key};
}

=head2 $req->set_encoding( $encoding )

Sets the character encoding used by the request's STDIN, STDOUT and STDERR
streams. This method may be called at any time to change the encoding in
effect, which will be used the next time C<read_stdin_line>, C<read_stdin>,
C<print_stdout> or C<print_stderr> are called. This encoding will remain in
effect until changed again. The encoding of a new request is determined by the
C<default_encoding> parameter of the containing C<Net::Async::FastCGI> object.
If the value C<undef> is passed, the encoding will be removed, and the above
methods will work directly on bytes instead of encoded strings.

=cut

sub set_encoding
{
   my $self = shift;
   my ( $encoding ) = @_;

   if( defined $encoding ) {
      my $codec = find_encoding( $encoding );
      defined $codec or croak "Unrecognised encoding '$encoding'";
      $self->{codec} = $codec;
   }
   else {
      undef $self->{codec};
   }
}

=head2 $line = $req->read_stdin_line

This method works similarly to the C<< <HANDLE> >> operator. If at least one
line of data is available then it is returned, including the linefeed, and
removed from the buffer. If not, then any remaining partial line is returned
and removed from the buffer. If no data is available any more, then C<undef>
is returned instead.

=cut

sub read_stdin_line
{
   my $self = shift;

   my $codec = $self->{codec};

   if( $self->{stdin} =~ s/^(.*[\r\n])// ) {
      return $codec ? $codec->decode( $1 ) : $1;
   }
   elsif( $self->{stdin} =~ s/^(.+)// ) {
      return $codec ? $codec->decode( $1 ) : $1;
   }
   else {
      return undef;
   }
}

=head2 $data = $req->read_stdin( $size )

This method works similarly to the C<read(HANDLE)> function. It returns the
next block of up to $size bytes from the STDIN buffer. If no data is available
any more, then C<undef> is returned instead. If $size is not defined, then it
will return all the available data.

=cut

sub read_stdin
{
   my $self = shift;
   my ( $size ) = @_;

   return undef unless length $self->{stdin};

   $size = length $self->{stdin} unless defined $size;

   my $codec = $self->{codec};

   # If $size is too big, substr() will cope
   my $bytes = substr( $self->{stdin}, 0, $size, "" );
   return $codec ? $codec->decode( $bytes ) : $bytes;
}

sub _print_stream
{
   my $self = shift;
   my ( $data, $stream ) = @_;

   while( length $data ) {
      # Send chunks of up to MAXRECORDDATA bytes at once
      my $chunk = substr( $data, 0, MAXRECORDDATA, "" );
      $self->write_record( { type => $stream, content => $chunk } );
   }
}

sub _flush_streams
{
   my $self = shift;

   if( length $self->{stdout} ) {
      $self->_print_stream( $self->{stdout}, FCGI_STDOUT );
      $self->{stdout} = "";
   }
   elsif( my $cb = $self->{stdout_cb} ) {
      $cb->();
   }

   if( length $self->{stderr} ) {
      $self->_print_stream( $self->{stderr}, FCGI_STDERR );
      $self->{stderr} = "";
   }
}

sub _needs_flush
{
   my $self = shift;
   return defined $self->{stdout_cb};
}

=head2 $req->print_stdout( $data )

This method appends the given data to the STDOUT stream of the FastCGI
request, sending it to the webserver to be sent to the client.

=cut

sub print_stdout
{
   my $self = shift;
   my ( $data ) = @_;

   my $codec = $self->{codec};

   $self->{stdout} .= $codec ? $codec->encode( $data ) : $data;

   $self->{conn}->_req_needs_flush( $self );
}

=head2 $req->print_stderr( $data )

This method appends the given data to the STDERR stream of the FastCGI
request, sending it to the webserver.

=cut

sub print_stderr
{
   my $self = shift;
   my ( $data ) = @_;

   my $codec = $self->{codec};

   $self->{used_stderr} = 1;
   $self->{stderr} .= $codec ? $codec->encode( $data ) : $data;

   $self->{conn}->_req_needs_flush( $self );
}

=head2 $req->stream_stdout_then_finish( $readfn, $exitcode )

This method installs a callback for streaming data to the STDOUT stream.
Whenever the output stream is otherwise-idle, the function will be called to
generate some more data to output. When this function returns C<undef> it
indicates the end of the stream, and the request will be finished with the
given exit code.

If this method is used, then care should be taken to ensure that the number of
bytes written to the server matches the number that was claimed in the
C<Content-Length>, if such was provided. This logic should be performed by the
containing application; C<Net::Async::FastCGI> will not track it.

=cut

sub stream_stdout_then_finish
{
   my $self = shift;
   my ( $readfn, $exitcode ) = @_;

   $self->{stdout_cb} = sub {
      my $data = $readfn->();

      if( defined $data ) {
         $self->print_stdout( $data );
      }
      else {
         delete $self->{stdout_cb};
         $self->finish( $exitcode );
      }
   };

   $self->{conn}->_req_needs_flush( $self );
}

=head2 $stdin = $req->stdin

Returns an IO handle representing the request's STDIN buffer. This may be read
from using the C<read> or C<readline> functions or the C<< <$stdin> >>
operator.

Note that this will be a tied IO handle, it will not be useable directly as an
OS-level filehandle.

=cut

sub stdin
{
   my $self = shift;

   return Net::Async::FastCGI::Request::TiedHandle->new(
      READ => sub { 
         $_[1] = $self->read_stdin( $_[2] );
         return defined $_[1] ? length $_[1] : 0;
      },
      READLINE => sub {
         return $self->read_stdin_line;
      },
   );
}

=head2 $stdout = $req->stdout

=head2 $stderr = $req->stderr

Returns an IO handle representing the request's STDOUT or STDERR streams
respectively. These may written to using C<print>, C<printf>, C<say>, etc..

Note that these will be tied IO handles, they will not be useable directly as
an OS-level filehandle.

=cut

sub _stdouterr
{
   my $self = shift;
   my ( $method ) = @_;

   return Net::Async::FastCGI::Request::TiedHandle->new(
      WRITE => sub { $self->$method( $_[1] ) },
   );
}

sub stdout
{
   return shift->_stdouterr( "print_stdout" );
}

sub stderr
{
   return shift->_stdouterr( "print_stderr" );
}

=head2 $req->finish( $exitcode )

When the request has been dealt with, this method should be called to indicate
to the webserver that it is finished. After calling this method, no more data
may be appended to the STDOUT stream. At some point after calling this method,
the request object will be removed from the containing C<Net::Async::FastCGI>
object, once all the buffered outbound data has been sent.

If present, C<$exitcode> should indicate the numeric status code to send to
the webserver. If absent, a value of C<0> is presumed.

=cut

sub finish
{
   my $self = shift;
   my ( $exitcode ) = @_;

   return if $self->is_aborted;

   $self->_flush_streams;

   # Signal the end of STDOUT
   $self->write_record( { type => FCGI_STDOUT, content => "" } );

   # Signal the end of STDERR if we used it
   $self->write_record( { type => FCGI_STDERR, content => "" } ) if $self->{used_stderr};

   $self->write_record( { type => FCGI_END_REQUEST, 
         content => build_end_request_body( $exitcode || 0, FCGI_REQUEST_COMPLETE )
   } );

   my $conn = $self->{conn};

   if( $self->{keepconn} ) {
      $conn->_removereq( $self->{reqid} );
   }
   else {
      $conn->close;
   }
}

=head2 $stdout = $req->stdout_with_close

Similar to the C<stdout> method, except that when the C<close> method is
called on the returned filehandle, the request will be finished by calling
C<finish>.

=cut

sub stdout_with_close
{
   my $self = shift;

   return Net::Async::FastCGI::Request::TiedHandle->new(
      WRITE => sub { $self->print_stdout( $_[1] ) },
      CLOSE => sub { $self->finish( 0 ) },
   );
}

sub _abort
{
   my $self = shift;
   $self->{aborted} = 1;

   my $conn = $self->{conn};
   $conn->_removereq( $self->{reqid} );

   delete $self->{stdout_cb};
}

=head2 $req->is_aborted

Returns true if the webserver has already closed the control connection. No
further work on this request is necessary, as it will be discarded.

It is not required to call this method; if the request is aborted then any
output will be discarded. It may however be useful to call just before
expensive operations, in case effort can be avoided if it would otherwise be
wasted.

=cut

sub is_aborted
{
   my $self = shift;
   return $self->{aborted};
}

=head1 HTTP::Request/Response Interface

The following pair of methods form an interface that allows the request to be
used as a source of L<HTTP::Request> objects, responding to them by sending
L<HTTP::Response> objects. This may be useful to fit it in to existing code
that already uses these.

=cut

=head2 $http_req = $req->as_http_request

Returns a new C<HTTP::Request> object that gives a reasonable approximation to
the request. Because the webserver has translated the original HTTP request
into FastCGI parameters, this may not be a perfect recreation of the request
as received by the webserver.

=cut

sub as_http_request
{
   my $self = shift;

   require HTTP::Request;

   my $params = $self->params;

   my $method = $params->{REQUEST_METHOD} || "GET";

   my $authority = 
      ( $params->{HTTP_HOST} || $params->{SERVER_NAME} || "" ) . ":" .
      ( $params->{SERVER_PORT} || "80" );

   my $path = join "", grep defined && length,
      $params->{SCRIPT_NAME},
      $params->{PATH_INFO};
   $path = "/" if !length $path;

   $path .= "?$params->{QUERY_STRING}" if length( $params->{QUERY_STRING} || "" );

   my $uri = URI->new( "http://$authority$path" )->canonical;

   my @headers;

   # Content-Type and Content-Length come specially
   push @headers, "Content-Type"   => $params->{CONTENT_TYPE}
      if exists $params->{CONTENT_TYPE};

   push @headers, "Content-Length" => $params->{CONTENT_LENGTH}
      if exists $params->{CONTENT_LENGTH};

   # Pull all the HTTP_FOO parameters as headers. These will be in all-caps
   # and use _ for word separators, but HTTP::Headers can cope
   foreach ( keys %$params ) {
      m/^HTTP_(.*)$/ and push @headers, $1 => $params->{$_};
   }

   my $content = $self->{stdin};

   return HTTP::Request->new( $method, $uri, \@headers, $content );
}

=head2 $req->send_http_response( $resp )

Sends the given C<HTTP::Response> object as the response to this request. The
status, headers and content are all written out to the request's STDOUT stream
and then the request is finished with 0 as the exit code.

=cut

sub send_http_response
{
   my $self = shift;
   my ( $resp ) = @_;

   # (Fast)CGI suggests this is the way to report the status
   $resp->header( Status => $resp->code );

   my $topline = $resp->protocol . " " . $resp->status_line;

   $self->print_stdout( $topline . $CRLF );
   $self->print_stdout( $resp->headers_as_string( $CRLF ) );

   $self->print_stdout( $CRLF );

   $self->print_stdout( $resp->content );
   $self->finish( 0 );
}

package # hide from CPAN
   Net::Async::FastCGI::Request::TiedHandle;
use base qw( Tie::Handle );

use Symbol qw( gensym );

sub new
{
   my $class = shift;

   my $handle = gensym;
   tie *$handle, $class, @_;

   return $handle;
}

sub TIEHANDLE
{
   my $class = shift;
   return bless { @_ }, $class;
}

sub CLOSE    { shift->{CLOSE}->( @_ ) }
sub READ     { shift->{READ}->( @_ ) }
sub READLINE { shift->{READLINE}->( @_ ) }
sub WRITE    { shift->{WRITE}->( @_ ) }

=head1 EXAMPLES

=head2 Streaming A File

To serve contents of files on disk, it may be more efficient to use
C<stream_stdout_then_finish>:

 use Net::Async::FastCGI;
 use IO::Async::Loop;

 my $fcgi = Net::Async::FastCGI->new(
    on_request => sub {
       my ( $fcgi, $req ) = @_;

       open( my $file, "<", "/path/to/file" );
       $req->print_stdout( "Status: 200 OK\r\n" .
                           "Content-type: application/octet-stream\r\n" .
                           "\r\n" );

       $req->stream_stdout_then_finish(
          sub { read( $file, my $buffer, 8192 ) or return undef; return $buffer },
          0
       );
    }

 my $loop = IO::Async::Loop->new();

 $loop->add( $fcgi );

 $loop->loop_forever;

It may be more efficient again to instead use the C<X-Sendfile> feature of
certain webservers, which allows the webserver itself to serve the file
efficiently. See your webserver's documentation for more detail.

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/Net/Async/FastCGI/Protocol.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2010-2011 -- leonerd@leonerd.org.uk

package Net::Async::FastCGI::Protocol;

use strict;
use warnings;

use base qw( IO::Async::Protocol::Stream );

our $VERSION = '0.24';

use Net::FastCGI::Constant qw( 
   FCGI_UNKNOWN_TYPE
);

use Net::FastCGI::Protocol qw(
   parse_header
   build_record
   build_unknown_type_body
);

sub on_read
{
   my $self = shift;
   my ( $buffref, $handleclosed ) = @_;

   my $blen = length $$buffref;

   if( $handleclosed ) {
      # Abort
      my $fcgi = $self->{fcgi};
      $fcgi->remove_child( $self );
      return;
   }

   # Do we have a record header yet?
   return 0 unless( $blen >= 8 );

   # Excellent - parse it
   my ( $type, $reqid, $contentlen, $padlen ) = parse_header( $$buffref );

   # Do we have enough for a complete record?
   return 0 unless( $blen >= 8 + $contentlen + $padlen );

   substr( $$buffref, 0, 8, "" ); # Header

   my $rec = {
      type  => $type,
      reqid => $reqid,
      len   => $contentlen,
      plen  => $padlen,
   };

   $rec->{content} = substr( $$buffref, 0, $contentlen, "" );

   substr( $$buffref, 0, $rec->{plen}, "" ); # Padding

   if( $reqid == 0 ) {
      $self->on_mgmt_record( $type, $rec );
   }
   else {
      $self->on_record( $reqid, $rec );
   }

   return 1;
}

sub on_mgmt_record
{
   my $self = shift;
   my ( $type, $rec ) = @_;

   $self->write_record( { type => FCGI_UNKNOWN_TYPE, reqid => 0 }, build_unknown_type_body( $type ) );
}

sub write_record
{
   my $self = shift;
   my ( $rec, $content ) = @_;

   $self->write( build_record( $rec->{type}, $rec->{reqid}, $content ) );
}

0x55AA;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.3800
use ExtUtils::MakeMaker;
WriteMakefile
(
  'NAME' => 'Net::Async::FastCGI',
  'VERSION_FROM' => 'lib/Net/Async/FastCGI.pm',
  'PREREQ_PM' => {
                   'Encode' => 0,
                   'Exporter' => '5.57',
                   'HTTP::Request' => 0,
                   'HTTP::Response' => 0,
                   'IO::Async::Listener' => '0.35',
                   'IO::Async::Loop' => '0.16',
                   'IO::Async::Protocol::Stream' => '0.33',
                   'IO::Async::Test' => 0,
                   'Net::FastCGI::Constant' => '0.10',
                   'Net::FastCGI::Protocol' => '0.10',
                   'Test::HexString' => 0,
                   'Test::More' => 0,
                   'Test::Refcount' => 0,
                   'Tie::Handle' => 0
                 },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [],
  'PL_FILES' => {}
)
;
